### ISSUE TITLE
Circular dependency when loading GitHub app private keys from Google Secret Manager

### ISSUE THREAD

**Issue by glasser** [google-github-actions/auth#386]
### TL;DR

We want to be able to keep all our secrets in Google Secret Manager, not GitHub Workflow secrets. This includes private keys to GitHub apps that we want to use to check out and push code. We have Workload Identity Federation set up properly.

I'd expect to be able to do these steps:
- `google-github-actions/auth` to authenticate with WIF
- `google-github-actions/get-secretmanager-secrets` to fetch the GH App private key
- `actions/create-github-app-token` to turn this into a token
- `actions/checkout` to clone the repository using the token

However, because `google-github-actions/auth` wants to come after `actions/checkout`, this doesn't work.

### Expected behavior

For there to be a reasonable way to load a GitHub App private key from GSM and use it to clone the repository.

### Observed behavior

The series of steps above (as indicated in the `google-github-actions/auth` docs) leads to the inability to use GCP APIs later.

### Action YAML

```yaml
name: Update Config
on:
  workflow_dispatch:

jobs:
  update-everything:
    runs-on: ubuntu-latest
    permissions:
      id-token: write
    
    steps:
      - uses: 'google-github-actions/auth@v2'
        with:
          workload_identity_provider: 'projects/PROJECTNUM/locations/global/workloadIdentityPools/github-d8bck/providers/github-d8bck'
          service_account: SA@PROJECTNAME.iam.gserviceaccount.com
          project_id: PROJECTNAME
      
      - id: gsm-secrets
        uses: 'google-github-actions/get-secretmanager-secrets@v2'
        with:
          secrets: |-
            automation-app-private-key:PROJECT/SECRET
        
      - uses: actions/create-github-app-token@v1
        id: automation-app-token
        with:
          app-id: APPNUM
          private-key: ${{ steps.gsm-secrets.outputs.automation-app-private-key }}
          owner: ${{ github.repository_owner }}

      - uses: actions/checkout@v4
        with:
          token: ${{ steps.automation-app-token.outputs.token }}

      - uses: A-CUSTOM-ACTION-THAT-USES-GCP-APIS
```

### Log output

```text
Error: The file at /home/runner/work/apollo-argo/apollo-argo/gha-creds-ca140d7f4adcce01.json does not exist, or it is not a file. ENOENT: no such file or directory, lstat '/home/runner/work/apollo-argo/apollo-argo/gha-creds-ca140d7f4adcce01.json'
```

### Additional information

If I understand correctly, the only reason `checkout` needs to come first is that otherwise it will overwrite the credentials file. So we are trying to work around this by adding a step that reads the credentials file before we run `checkout` and another one that writes it back again afterwards. This is aided by the fact that this action has the credentials filename as an output, though it would be nice if the contents were also an output so we wouldn't have to read it back (if that's OK from a secret perspective)?  At the very least there could be some documentation about approaches to take if you have a real need to run this before `checkout`? (Also, you need to do something silly like write a dummy file to the workspace if you want to avoid the empty-workspace warning.)

**Comment by glasser** [comment]
By the way I do recognize that there are many issues about people wanting to put the credentials somewhere other than the workspace, and I respect that you're not going to change that. I would just like there to be *some* non-hacky way to do what seems to be a reasonable thing to do (check out the repository using a token derived from a GSM secret).  Even being able to do something like run `auth` again in a mode that just regenerates the credentials file without redoing the work would be nice... though actually I can't figure out a nice way to implement that.

Is the actual best answer here to just run `auth` a second time after checkout?

**Comment by sethvargo** [comment]
Checkout twice, using a different path for the second checkout:

```yaml
- uses: 'actions/checkout@v4'
- uses: 'google-github-actions/auth@v2'
- uses: 'google-github-actions/get-secretmanager-secrets@v2'
- uses: 'actions/checkout@v4'
  with:
    path: 'subdir'
```

You might also try setting `clean: false` on the second checkout and seeing if it stops the clobbering.

Finally, the empty workspace warning is just a warning. It can be ignored if you know you don't need to persist credentials.

**Comment by glasser** [comment]
Thanks for the advice. This does seem like it would be a fair amount slower than my hacky workaround, though I'll try to profile before making assumptions. Am I understanding correctly that the main thing the action uses the checkout for is a location to store the credentials file, or are there other problems I'd run into?

(And I understand that it's just a warning, which is why I feel comfortable suppressing it — I don't want all my colleagues to have to worry that it's something they need to pay attention to!)
